### ISSUE TITLE
Formatting Node labels (not the hovering label) Force-graph 2D

### ISSUE THREAD

**Issue by nalindeepan007** [vasturiano/react-force-graph#433]
@vasturiano 
Currently, node labels are displaying like this, with **nodeCanvasObject**
the text is spilling out of circular node:

![image](https://user-images.githubusercontent.com/22245183/228772598-6ad11754-7c61-4ec7-ac4c-a08cf77002d5.png)

How can I set node labels like this when the text is longer (or a full sentence), so that it can wrap inside the circular node...
![image](https://user-images.githubusercontent.com/22245183/228772723-838e9bc9-23c2-40a7-9114-486e4ba58eba.png)

Thanks

**Comment by psygo** [comment]
I would also very much like to know how to do this.

**Comment by vasturiano** [comment]
@psygo this question is kind of beyond the scope of this package. Since you're defining your own custom node (via `nodeCanvasObject`) you're really in full control on how to write the text in your custom node.

But in any case, you can find some examples around on how to wrap text inside a circle in a Canvas environment. For example:
https://observablehq.com/@mbostock/fit-text-to-circle

**Comment by psygo** [comment]
Thank you for the helpful answer, @vasturiano !

I think that now probably at least half the problem should hopefully be solved with that link.

But do you know how to draw the text on top of the nodes? For some reason, adding `     `ctx.fillText("hello", node.x!, node.y!)` to your [Highligh nodes/links](https://github.com/vasturiano/react-force-graph/blob/master/example/highlight/index.html) example keeps drawing the text behind the nodes. I've also tried variations with `nodeCanvasObjectMode`, with no success.

![image](https://github.com/vasturiano/react-force-graph/assets/11823725/8af2cd07-c7da-4745-8da6-e0389f2d87e4)

**Comment by psygo** [comment]
Strange, cause I did try various combinations of `before`, `after`, and `replace` for `nodeCanvasObjectMode` in that example, and none of them worked to draw on top of the node.

**Comment by vasturiano** [comment]
Do you have a repro example?

**Comment by psygo** [comment]
The pictures I was providing in previous comments were very basically the [Highlight nodes/links](https://github.com/vasturiano/react-force-graph/blob/master/example/highlight/index.html) example you provided in the documentation. The only thing I did was add `ctx.fillText("hello", node.x!, node.y!)` to `paintRing`:

```js
...

const paintRing = useCallback(
  (node, ctx) => {
    // add ring just for highlighted nodes
    ctx.beginPath();
    ctx.arc(
      node.x,
      node.y,
      NODE_R * 1.4,
      0,
      2 * Math.PI,
      false
    );
    ctx.fillStyle = node === hoverNode ? "red" : "orange";
    ctx.fill();

    // Doesn't seem to matter if I put it before or after `ctx.fill()`
    ctx.fillText("hello", node.x!, node.y!)  
  },
  [hoverNode]
);

return (
  <ForceGraph2D
    graphData={data}
    nodeRelSize={NODE_R}
    autoPauseRedraw={false}
    linkWidth={(link) =>
      highlightLinks.has(link) ? 5 : 1
    }
    linkDirectionalParticles={4}
    linkDirectionalParticleWidth={(link) =>
      highlightLinks.has(link) ? 4 : 0
    }
    // Doesn't seem to work to draw text on top of the node
    // nodeCanvasObjectMode="after"
    nodeCanvasObjectMode={(node) =>
      highlightNodes.has(node) ? "before" : undefined
    }
    nodeCanvasObject={paintRing}
    onNodeHover={handleNodeHover}
    onLinkHover={handleLinkHover}
  />
);
```

![Screenshot_2023-11-12-21-03-07_5120x1080](https://github.com/vasturiano/react-force-graph/assets/11823725/c53be3e0-c3a2-430d-a8d0-2dd9e52fb72b)

**Comment by heldersepu** [comment]
@psygo see my second response in StackOverflow ...

**Comment by heldersepu** [comment]
Here is a more complete test:
https://raw.githack.com/heldersepu/hs-scripts/master/HTML/ForceGraph2D.html

Actual code:
https://github.com/heldersepu/hs-scripts/blob/master/HTML/ForceGraph2D.html

**Comment by heldersepu** [comment]
To address the original question from @nalindeepan007 ... wrapping text inside the circular node is not an easy task, you have to come up with an algorithm that suits your needs, there is no magic bullet for that

**Comment by psygo** [comment]
Thank you for the helpful answer, @heldersepu ! I had actually tried much of what you've shown here, but somehow it didn't work, I don't know why. Thank you for showing that it does work.

> To address the original question from @nalindeepan007 ... wrapping text inside the circular node is not an easy task,

I haven't had the time to test it, but doesn't the link [Mike Bostock's tutorial on how to fit text to a circle]() provided by @vasturiano solve that issue?

**Comment by heldersepu** [comment]
@psygo Yes, that ("Fit Text to Circle") is just an example of how to fit text into an circle in SVG, but some of that logic can still apply to any other context, you just have to put the time an fill in the gaps...

@psygo I think you can this issue as out of scope

**Comment by psygo** [comment]
Somebody provided an answer as to how to fit the text to a circle with `<canvas>` [here](https://stackoverflow.com/a/77603059/4756173). That piece of code is a bit difficult to understand, and when I used it in React, it behaved a bit weirdly on the first render, but it seems to work fine. Anyways, I'm gonna leave it here if anyone finds it useful.
